doc/examples/uuid.rst

```diff
@@ -0,0 +1,290 @@
+cross-language compatibility.
+
+.. attention:: Applications that do not share a MongoDB deployment with
+   any other application and that have never stored UUIDs in MongoDB
+   should use the ``standard`` UUID representation for cross-language
+   compatibility. See :ref:`configuring-uuid-representation` for details
+   on how to configure the :class:`~bson.binary.UuidRepresentation`.
+
+.. _example-legacy-uuid:
+
+Legacy Handling of UUID Data
+----------------------------
+
+Historically, MongoDB Drivers have used different byte-ordering
+while serializing UUID types to :class:`~bson.binary.Binary`.
+Consider, for instance, a UUID with the following canonical textual
+representation::
+
+  00112233-4455-6677-8899-aabbccddeeff
+
+This UUID would historically be serialized by the Python driver as::
+
+  00112233-4455-6677-8899-aabbccddeeff
+
+The same UUID would historically be serialized by the C# driver as::
+
+  33221100-5544-7766-8899-aabbccddeeff
+
+Finally, the same UUID would historically be serialized by the Java driver as::
+
+  77665544-3322-1100-ffee-ddccbbaa9988
+
+.. note:: For in-depth information about the the byte-order historically
+   used by different drivers, see the `Handling of Native UUID Types
+   Specification
+   <https://github.com/mongodb/specifications/blob/master/source/uuid.rst>`_.
+
+This difference in the byte-order of UUIDs encoded by different drivers
+resulted in highly unintuitive behavior in deployments where more than
+one of these drivers was in use. As example, consider the following situation:
+
+* Application ``M`` written in C# generates a UUID and uses it as the ``_id``
+  of a document that it proceeds to insert into the ``uuid_test`` collection of
+  the ``example_db`` database. Let's assume that the canonical textual
+  representation of the generated UUID is::
+
+    00112233-4455-6677-8899-aabbccddeeff
+
+* Application ``N`` written in Python attempts to ``find`` the document
+  written by application ``M`` in the following manner::
+
+    from uuid import UUID
+    collection = client.example_db.uuid_test
+    results = collection.find({'_id': UUID('00112233-4455-6677-8899-aabbccddeeff')})
+
+  In this instance, the ``results`` cursor will never contain the document that
+  was inserted by application ``M`` in the previous step. This is because of
+  the different byte-order used by the C# driver for representing UUIDs as
+  BSON Binary. The following query, on the other hand, will successfully find
+  this document::
+
+    results = collection.find({'_id': UUID('33221100-5544-7766-8899-aabbccddeeff')})
+
+As this example demonstrates, differing byte-order can hamper
+interoperability between applications that use different drivers. To workaround
+this problem, users can configure their ``MongoClient`` with the appropriate
+:class:`~bson.binary.UuidRepresentation`.
+
+.. _configuring-uuid-representation:
+
+Configuring a UUID Representation
```

Comment:
Somewhere in here you should talk about round tripping data, and you might want to warn about it. It's easy for an application to read and update a record accidentally changing the byte order of the object in MongoDB.

Reply:
@prashantmital, an example of changing byte order would be round tripping a Binary subtype 4 with uuid representation C# or Java. The subtype 4 would be decoded to a native UUID and then encoded as Binary subtype 3 with **different** bytes.

Reply:
As clarified on Slack, the bytes are not different, but the subtype is. This is reflected in the new example.

Reply:
No, the bytes would also change:
```
>>> u = uuid.UUID('00112233445566778899AABBCCDDEEFF')
>>> b_standard = bson.encode({'u': u}, codec_options=CodecOptions(uuid_representation=UuidRepresentation.STANDARD))
>>> b_standard
b'\x1d\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x04\x00\x11"3DUfw\x88\x99\xaa\xbb\xcc\xdd\xee\xff\x00'
>>> decoded = bson.decode(b_standard, codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> decoded
{'u': UUID('00112233-4455-6677-8899-aabbccddeeff')}
>>> u
UUID('00112233-4455-6677-8899-aabbccddeeff')
>>> # The UUID's subtype AND data change here:
>>> b_legacy = bson.encode(decoded, codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> b_legacy
b'\x1d\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x03wfUD3"\x11\x00\xff\xee\xdd\xcc\xbb\xaa\x99\x88\x00'
>>> b_standard
b'\x1d\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x04\x00\x11"3DUfw\x88\x99\xaa\xbb\xcc\xdd\xee\xff\x00'
>>> decoded = bson.decode(b_legacy, codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> decoded
{'u': UUID('00112233-4455-6677-8899-aabbccddeeff')}
>>> decoded = bson.decode(b_legacy, codec_options=CodecOptions(uuid_representation=UuidRepresentation.STANDARD))
>>> decoded
{'u': UUID('77665544-3322-1100-ffee-ddccbbaa9988')}
```

Reply:
I am don't follow this completely. Will need to discuss.

Reply:
Here's the same example written from a different perspective. App `A` inserts a document with STANDARD. App `B` find and replaces the same document with JAVA_LEGACY.
```python
>>> coll_standard = client.t.get_collection('t', codec_options=CodecOptions(uuid_representation=UuidRepresentation.STANDARD))
>>> coll_java = client.t.get_collection('t', codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> coll_raw = client.t.get_collection('t', codec_options=CodecOptions(document_class=RawBSONDocument))
>>> u = UUID('00112233-4455-6677-8899-aabbccddeeff')
>>> coll_standard.insert_one({'_id': 1, 'u': u})
<pymongo.results.InsertOneResult object at 0x7fe15e453500>
>>> # Raw bytes on the server:
>>> coll_raw.find_one({'_id':1})
RawBSONDocument(b'&\x00\x00\x00\x10_id\x00\x01\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x04\x00\x11"3DUfw\x88\x99\xaa\xbb\xcc\xdd\xee\xff\x00', ...)
>>> # Find/replace the document from a legacy Java app:
>>> doc = coll_java.find_one({'_id': 1})
>>> coll_java.replace_one({'_id': 1}, doc)
<pymongo.results.UpdateResult object at 0x7fe15e453b00>
>>> # Raw bytes on the server have changed:
>>> coll_raw.find_one({'_id':1})
RawBSONDocument(b'&\x00\x00\x00\x10_id\x00\x01\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x03wfUD3"\x11\x00\xff\xee\xdd\xcc\xbb\xaa\x99\x88\x00', ...)
```

Reply:
Added an example for this.